shared_model/builders/protobuf/transaction.hpp

```diff
@@ -46,131 +47,173 @@ namespace shared_model {
       };
 
       template <int s>
-      using NextBuilder = TemplateTransactionBuilder<S | (1 << s)>;
+      using NextBuilder = TemplateTransactionBuilder<S | (1 << s), SV>;
 
-      iroha::protocol::Transaction transaction_;
+      using ProtoTx = iroha::protocol::Transaction;
+      using ProtoCommand = iroha::protocol::Command;
 
       template <int Sp>
-      TemplateTransactionBuilder(const TemplateTransactionBuilder<Sp> &o,
-                                 SV&& stateless_validator = SV())
+      TemplateTransactionBuilder(const TemplateTransactionBuilder<Sp> &o)
           : transaction_(o.transaction_),
-            stateless_validator_(std::forward<SV>(stateless_validator)) {}
+            stateless_validator_(o.stateless_validator_) {}
+
+      TemplateTransactionBuilder(const ProtoTx &tx,
+                                 const SV &stateless_validator)
+          : transaction_(tx), stateless_validator_(stateless_validator) {}
+
+      /**
+       * Make transformation on copied content
+       * @tparam Transformation - callable type for changing the copy
+       * @param f - transform function for proto object
+       * @return new builder with updated state
+       */
+      template <int Fields, typename Transformation>
+      NextBuilder<Fields> transform(Transformation t) const {
+        auto copy = transaction_;
+        t(copy);
+        return {copy, stateless_validator_};
+      }
 
-      SV stateless_validator_;
+      /**
+       * Make add command transformation on copied object
+       * @tparam Transformation - callable type for changing command
+       * @param f - transform function for proto command
+       * @return new builder with added command
+       */
+      template <typename Transformation>
+      NextBuilder<Command> addCommand(Transformation t) const {
+        auto copy = transaction_;
+        t(copy.mutable_payload()->add_commands());
+        return {copy, stateless_validator_};
+      }
 
      public:
       TemplateTransactionBuilder() = default;
 
       NextBuilder<CreatorAccountId> creatorAccountId(
-          const interface::types::AccountIdType &account_id) {
-        transaction_.mutable_payload()->set_creator_account_id(account_id);
-        return *this;
+          const interface::types::AccountIdType &account_id) const {
+        return transform<CreatorAccountId>([&](auto &tx) {
+          tx.mutable_payload()->set_creator_account_id(account_id);
+        });
       }
 
-      NextBuilder<TxCounter> txCounter(Transaction::TxCounterType tx_counter) {
-        transaction_.mutable_payload()->set_tx_counter(tx_counter);
-        return *this;
+      NextBuilder<TxCounter> txCounter(
+          Transaction::TxCounterType tx_counter) const {
+        return transform<CreatorAccountId>([&](auto &tx) {
```

Comment:
Should be `transform<TxCounter>`. Same with `CreatedTime` below. Maybe change return type to `auto` to write this only once?